src/interface/comm.rs

```diff
@@ -366,3 +369,95 @@ impl Drop for Socket {
 pub fn getifaddrs_from_file() -> String {
     read_to_string(NET_DEV_FILENAME).expect("No net_devices file present!").to_owned()
 }
+
+// Implementations of select related FD_SET structure
+pub struct FdSet(libc::fd_set);
+
+impl FdSet {
+    pub fn new() -> FdSet {
+        unsafe {
+            let mut raw_fd_set = std::mem::MaybeUninit::<libc::fd_set>::uninit();
+            libc::FD_ZERO(raw_fd_set.as_mut_ptr());
+            FdSet(raw_fd_set.assume_init())
+        }
+    }
+
+    pub fn new_from_ptr(raw_fdset_ptr: *const libc::fd_set) -> &'static mut FdSet {
+        unsafe {
+            &mut *(raw_fdset_ptr as *mut FdSet)
+        }
+    }
+
+    // copy the src FdSet into self
+    pub fn copy_from(&mut self, src_fds: &FdSet) {
+        unsafe {
+            std::ptr::copy_nonoverlapping(&src_fds.0 as *const libc::fd_set, &mut self.0 as *mut libc::fd_set, 1);
+        }
+    }
+
+    // turn off the fd bit in fd_set (currently only used by the tests)
+    #[allow(dead_code)]
+    pub fn clear(&mut self, fd: RawFd) {
+        unsafe { libc::FD_CLR(fd, &mut self.0) }
+    }
+
+    // turn on the fd bit in fd_set
+    pub fn set(&mut self, fd: RawFd) {
+        unsafe { libc::FD_SET(fd, &mut self.0) }
+    }
+
+    // return true if the bit for fd is set, false otherwise
+    pub fn is_set(&self, fd: RawFd) -> bool {
+        unsafe { libc::FD_ISSET(fd, &self.0) }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        let fd_array: &[u8] = unsafe {
+            std::slice::from_raw_parts(&self.0 as *const _ as *const u8, mem::size_of::<libc::fd_set>())
+        };
+        fd_array.iter().all(|&byte| byte == 0)
+    }
+
+    // for each fd, if kernel_fds turned it on, then self will turn the corresponding tranlated fd on
+    pub fn set_from_kernelfds_and_translate(&mut self, kernel_fds: &FdSet, nfds: i32, rawfd_lindfd_tuples: &Vec<(i32, i32)>) {
```

Comment:
theres a lot of nesting here, maybe condense some of this to one liners for readability

Reply:
Following Nick's suggestion, the nested loop can be modified as something like this:
```
use std::collections::HashMap;

pub fn set_from_kernelfds_and_translate(&mut self, kernel_fds: &FdSet, nfds: i32, rawfd_lindfd_tuples: &Vec<(i32, i32)>) {
    // Create a HashMap from the rawfd_lindfd_tuples for O(1) look-up
    let fd_map: HashMap<i32, i32> = rawfd_lindfd_tuples.iter().cloned().collect();
    for fd in 0..nfds {
        if kernel_fds.is_set(fd) {
            if let Some(lindfd) = fd_map.get(&fd) {
                self.set(*lindfd);
            }
        }
    }
}
```

Reply:
this is probably not a great idea because I believe creating a hashmap is slow, so even though the algorithmic complexity may be better N should never be large enough where it surpasses that allocation cost.